[flang][runtime] Add ACCESS library procedure

tblah opened the pull request:
This is a GNU extension:
https://gcc.gnu.org/onlinedocs/gfortran/ACCESS.html

Used in SALMON:
https://salmon-tddft.jp/download.html

Unfortunately the intrinsic takes a file path to operate on so there isn't an easy way to make the test robust. The unit test expects to be able to create, set read write and execute permissions, and delete files called
  `std::filesystem::temp_directory_path() / <test_name>.<pid>`

The test will fail if a file already exists with that name.

I have not implemented the intrinsic on Windows because this is wrapping a POSIX system call and Windows doesn't support all of the permission bits tested by the intrinsic. I don't have a Windows machine easily available to check if Gfortran implements this intrinsic on Windows.

tblah committed:
[flang][runtime] Add ACCESS intrinsic

This is a GNU extension:
https://gcc.gnu.org/onlinedocs/gfortran/ACCESS.html

Used in SALMON:
https://salmon-tddft.jp/download.html

Unfortunately the intrinsic takes a file path to operate on so there
isn't an easy way to make the test robust. The unit test expects to be
able to create, set read write and execute permissions, and delete
files called
  std::filesystem::temp_directory_path() / <test_name>.<pid>

The test will fail if a file already exists with that name.

I have not implemented the intrinsic on Windows because this is wrapping
a POSIX system call and Windows doesn't support all of the permission
bits tested by the intrinsic. I don't have a Windows machine easily
available to check if Gfortran implements this intrinsic on Windows.

klausler reviewed (commented):
This looks like you've added `ACCESS` as a library procedure, not as an intrinsic function with support in semantics, lowering, and I/O runtime.

tblah commented:
Thanks for taking a look

> This looks like you've added `ACCESS` as a library procedure, not as an intrinsic function with support in semantics, lowering, and I/O runtime.

Yes. I was trying to keep things simple. This is similar to FDATE: 959a430a8d

klausler reviewed (approved):
SGTM, and LGTM.  Please change the commit title to prevent any future confusion.

kiranchandramohan commented on flang/runtime/extensions.cpp:
```diff
@@ -138,5 +139,77 @@ void RTNAME(Sleep)(std::int64_t seconds) {
+#ifndef _WIN32
+std::int64_t FORTRAN_PROCEDURE_NAME(access)(const char *name,
+    std::int64_t nameLength, const char *mode, std::int64_t modeLength) {
+  std::int64_t ret = -1;
```
Nit: Use Braced initialization here and elsewhere in this file.

tblah committed:
Add missing guard to prevent adding code to a Windows build

tblah committed:
Use braced initialization

I didn't apply braced initialization to very common C programming
patterns in the unittest because when something is extremely common, I
think it increases cognitive load when reading. For example
  int fd = open(...);
vs
  int fd{open(...)};

I did change all uses in the runtime implementation as requested in code
review.

yiwu0b11 reviewed (approved):
LGTM, thanks for adding this.
